### DISCUSSION TITLE
Anyone got Firebase working?

### DISCUSSION THREAD

**Discussion by williambuchanan2** [dotnet/maui#18023] [General] [answered]
I am using Firebase in my Maui app for push notification. It works on Android but I am having a problem with WillPresentNotification not being called on iOS. 

I have made a new test app in iOS and it works fine but for some reason it doesn't work in my main app. I can't see what I have missed. Everything looks the same.

Any ideas? What could cause that not to fire?

Alternatively, are there any other push notification options that are easier to get working?

Thanks

**Comment by jcsnider** [comment, marked as answer]
The latest ShinyLib 3.X.X betas are working well for me in the .Net 8 RCs.

**Reply by williambuchanan2** [reply]
I had looked at that but it seemed like a big package for what I needed, and I wasn't sure how easy it looked to get up and running. In the end I got Firebase working. It turned out there was something else consuming the message notification so once we discovered that and removed it the messages came in.

Looks like there aren't really any nice clean simple solutions to this at the moment.

**Reply by alzubitariq** [reply]
I have the same issue , @williambuchanan2 what was that ? "there was something else consuming the message notification" how did you fix the issue ?

**Reply by williambuchanan2** [reply]
One of the devs had installed a plugin for local notifications and it was consuming the push noti event. As soon as we removed it the event started triggering. The only way I found it was by removing things 1 by 1 until it started working!

**Reply by thomasgalliker** [reply]
Side note: The goal of „the new plugin“ is to have local and push notification working side by side. We use local notifications too and that was exaclty our problem too.
👍🏻

**Comment by TedMobile** [comment]
Hi Cannot get firebase working on iOS -**messaging.sharedinstance is always null** . does anyone have a sample using push notification with iOS? Happy to start using Shiny but could not get it working and probably missing something obvious. Anything really to get me going

many thanks

**Reply by williambuchanan2** [reply]
Even though there isn't much to it, it is really fiddly to get working. Below are the key snippets of my code which I have in place (note - it's not a complete example but will hopefully help).

Don't forget you also need to download the plist file GoogleService-Info.plist from Firebase and have this in the project file as a bundled resource:

		<BundleResource Include="Platforms\iOS\GoogleService-Info.plist" Link="GoogleService-Info.plist" />

You might need to enter that manually into the csproj, and also I noticed that the file might need to be unblocked after you download it because VS can't read it (Windows Explorer -> right click -> Properties -> Unblock).

The code below has the 3 things you need:
1. User taps the notification from the notification area.
2. New notification arrives on the device.
3. New registration token issued.

Hope this helps in some way....
````
   public class UserNotificationCenterDelegate : UNUserNotificationCenterDelegate
   {
       /// Fires when the user taps a notification from the device notification area
       public override void DidReceiveNotificationResponse(UNUserNotificationCenter center, UNNotificationResponse response, Action completionHandler)
       {
               var jsonContent = response.Notification.Request.Content.UserInfo.ValueForKey(new NSString(@"Action"));
              // Code that you want to call when the noti is tapped
              ....
       }

        /// Fires when a new notification arrives
       public override void WillPresentNotification(UNUserNotificationCenter center, UNNotification notification, Action<UNNotificationPresentationOptions> completionHandler)
       {
               var jsonContent = notification.Request.Content.UserInfo.ValueForKey(new NSString(@"Action"));
               completionHandler(UNNotificationPresentationOptions.None);
              // Code that you want to call when a new noti arrives
              ....
       }
   }


    public class AppDelegate : MauiUIApplicationDelegate, IMessagingDelegate
    {
        public override bool FinishedLaunching(UIApplication application, NSDictionary launchOptions)
        {
           var result = base.FinishedLaunching(application, launchOptions);
           InitPushNotification(new UserNotificationCenterDelegate ());
           return result;
        }

        public void InitPushNotification(IUNUserNotificationCenterDelegate userNotificationCenterDelegate)
        {
            Firebase.Core.App.Configure();

            if (UIDevice.CurrentDevice.CheckSystemVersion(10, 0))
            {
                var authOption = UNAuthorizationOptions.Alert | UNAuthorizationOptions.Badge | UNAuthorizationOptions.Sound;
                UNUserNotificationCenter.Current.RequestAuthorization(authOption, (granted, error) =>
                {
                });

                UNUserNotificationCenter.Current.Delegate = userNotificationCenterDelegate;

                Messaging.SharedInstance.AutoInitEnabled = true;
                Messaging.SharedInstance.Delegate = _messageDelegate;
            }
            UIApplication.SharedApplication.RegisterForRemoteNotifications();
        }

        /// Fires when a new device token is issued by Firebase
        [Export("messaging:didReceiveRegistrationToken:")]
        public void DidReceiveRegistrationToken(Messaging message, string regToken)
        {
            // Record the regToken somewhere to be used when sending the noti
        }
   }
````

**Comment by TedMobile** [comment]
Hi ,
Many thanks for your reply
I have done all the steps but the unblock of the file and when I go to win11 winexplorer properties - general tab that button is not there for me.
I still get this sharedinstance is null which is so annoying and I have even deployed and downloaded the app from appcenter but I still get  sharedinstance is null..  and my code is very similar to yours with    Firebase.Core.App.Configure etc.. but as soon as it hits  Messaging.SharedInstance.AutoInitEnabled = true;  it crashes ..

what Am I missing?

**Reply by williambuchanan2** [reply]
For the unblock button - if the button doesn't appear then it means the file is ok, so shouldn't be a problem. This is what it looks like.

![image](https://github.com/dotnet/maui/assets/32165641/18c07c35-d930-4c92-97b9-9af066133939)

**Reply by williambuchanan2** [reply]
Have you got IMessagingDelegate on AppDelegate?

![image](https://github.com/dotnet/maui/assets/32165641/5f690cb8-b723-4f4f-83d3-ddec8fbed234)

**Reply by TedMobile** [reply]
Oh let me try that tomorrow as I off today I don’t think I am using that as I read there was no longer a need but now I am so desperate to get something running that I will try anything

**Reply by TedMobile** [reply]
Of course no reply from Microsoft on the many issues raised in their repo related to this. It seems so many people got stuck on this one

**Reply by TedMobile** [reply]
Of course no reply from Microsoft on the many issues raised in their repo related to this. It seems so many people got stuck on this one

**Reply by williambuchanan2** [reply]
You won't get any action from Microsoft for months, if ever. They will ask you to provide a reproducable demo. If you do they will just throw it in the backlog. If you don't they will just close it. Either way it won't be looked at for a long time.

**Comment by thomasgalliker** [comment]
Not sure if it solves your problem, but, we're using this plugin: https://github.com/thomasgalliker/Plugin.FirebasePushNotifications
It's about to be refactored entirely... and there is no documentation yet. The plugin works in our case - but we're stuck with .NET 7 MAUI. Not sure if it fits your needs. Let me know if it does or doesnt.

**Reply by TedMobile** [reply]
Oh I will have a go tomorrow as I am out of office today does it have a sample app to quickly get it up and running. I am using visual studio and windows ? I suppose iOS works for u

**Reply by thomasgalliker** [reply]
Yes. There is a small sample app in the git repo but its far away from a production app. Have a look at the refactoring branch, https://github.com/thomasgalliker/Plugin.FirebasePushNotifications/tree/refactoring-cleanup

Lets continue the discussion there. I will be back for the good on Monday.

**Reply by williambuchanan2** [reply]
Wish I had seen this a couple of weeks ago. It took 2 of us over a week to get it working nicely on both platforms. There isn't really a good option for this at the moment so if you can get this out there I imagine it will be welcomed by the community!

**Reply by AlleSchonWeg** [reply]
Is this a fork/copy from https://github.com/CrossGeeks/FirebasePushNotificationPlugin ?
Here is another try to bring the plugin to maui: https://github.com/Gekidoku/BetterFireBaseNotificationsPlugin

**Reply by TedMobile** [reply]
Hi all , it must be something wrong with my machine or my sample.. 
@williambuchanan2  - you are absolutely right Microsoft will never fix that but we are really stuck then!!! 
Also for the life of me I cannot get it work
My class uses
`public partial class NotificationService : UNUserNotificationCenterDelegate, INotificationService, IMessagingDelegate`   INotificationService is my interface...
Why do I get this "SharedInstance is null - I think the GoogleService-Info.plist is not read may be but in the project I have put

	<ItemGroup>
		<GoogleServicesJson Include="Platforms\Android\google-services.json" />
	</ItemGroup>
	<ItemGroup>
		<BundleResource Include="Platforms\iOS\GoogleService-Info.plist" Link="GoogleService-Info.plist" />
	</ItemGroup>

@thomasgalliker  I tried your plugin and I get the same issues it crashes here.. and again SharedInstance is null.
`        Firebase.CloudMessaging.Messaging.SharedInstance.AutoInitEnabled = defaultOptions.AutoInitEnabled;`


Why does it work for you guys - What Am I missing? may be I am doing some basic stuff wrong but then also these other people :

**some issues I found all around this.** 
https://github.com/xamarin/xamarin-macios/issues/19182
https://github.com/xamarin/GoogleApisForiOSComponents/issues/621
https://github.com/xamarin/GoogleApisForiOSComponents/issues/486
https://stackoverflow.com/questions/76677155/messaging-sharedinstance-delegate-this-is-null-in-xamarin-ios

also - slightly worrying there is this thread here where this person says that the app got rejected because firebase bitcode, I think internally that nuget uses firebase.core
**App store denied** 
https://github.com/dotnet/maui/issues/18162

this all thing is driving me insane to get it working in iOS

**Reply by AlleSchonWeg** [reply]
@TedMobile You are using HotRestart (connecting a iPhone on a windows machine without a mac)?

**Reply by thomasgalliker** [reply]
I will come back to this @TedMobile!
The plugin is a fork of the original push plugin, yes. Nobody seemed to care about the CrossGeeks plugin and above all it had (still has) some really nasty design flaws in the code. Thats why I started the refactoring branch. There is still a LOT to improve, all these statis variables need to get replaced. 
i also added event queues to catch notifications and store them in queues - in case your app is not yet ready to process them. 
I hope I get some more time to make a brilliant and easy to use piece of nuget 😎

**Comment by gabsamples6** [comment]
@AlleSchonWeg I get exactly the same behavior and I use HotRestart (connecting a iPhone on a windows machine without a mac) but just noticed that a lead/dev on the MacOS team has confirmed the issue https://github.com/xamarin/xamarin-macios/issues/19182

**Reply by AlleSchonWeg** [reply]
Ok,
with HotRestart this will not work. It's because of known limitations: https://learn.microsoft.com/en-us/dotnet/maui/ios/hot-restart#limitations

I had the same problem with Xamarin Forms a few month ago. The cause is that Firebase uses static librarys and this is not supported by HotRestart.
At the end i use a mac to create an ipa with ad hoc distribution profile. This ipa can be installed on the iPhone. Then it works.
To make the rest working with HotRestart i uses the  HOTRESTART preprocessor symbol:
```C#
#if !HOTRESTART
  // Code here won't be executed when debugging with hot restart
			FirebasePushNotificationManager.Initialize(options, false);
			Messaging.SharedInstance.AutoInitEnabled = true;
			Messaging.SharedInstance.Delegate = CrossFirebasePushNotification.Current as IMessagingDelegate;
			CrossFirebasePushNotification.Current.OnNotificationError += Current_OnNotificationError;
#endif
```
With ios simulator it's also not possible receive notifications from firebase.
Fazit: You need a mac to create an ipa and distribute it via ad hoc or testflight.

**Reply by AlleSchonWeg** [reply]
Just for completeness.
Starting with Xcode 14 and iOS 16 apple enabled push support for simulators. Also firebase supports the scenario: https://github.com/firebase/firebase-ios-sdk/pull/10503

But with XF (i think MAUI too) the firebase ios sdk bindings are VEEEEERY old. (https://github.com/xamarin/GoogleApisForiOSComponents) Android are up to date

**Comment by TedMobile** [comment]
@AlleSchonWeg  just seen this - interesting!!
I honestly was not aware of the "Hotrestart preprocessor" but what I did I built and deployed the app using azure devops and deployed to appcenter(via azure devops) downloaded the app on the iPhone but still same error.. makes no sense. 

- We use  windows mainly for dev with iphone 
-   macmini for running against simulators via windows PC.. 
-  Azure devops to deploy to appcenter (used only for test distribution) and upload to testflight and android testing track
-
 we had so many issues with VS for mac so we stuck to windows..   

I had a glance at the code base and indeed is so old - android just works. 

**in conclusion**
Should this have worked when deployed as hotrestart is out of the picture?

Also what about using APNS directly has anyone done that  I mean (no xamarin.firebase)

**Reply by AlleSchonWeg** [reply]
We never used azure devops and appcenter distribution. So i have no idea how it works.

I only used VS for mac to create an ad hoc ipa or release.
I debugged the push notifications via `Console.Writline()`. You could try this also. Connect the iPhone to the windows pc and open the Device Log window in VS. Search for firebase, push, your bundle ident or your application name. Perhaps you find some info why firebase can't init.

You should also check if push notifications capability is enabled for you app id at https://developer.apple.com/

**Reply by williambuchanan2** [reply]
Good to hear you found the reason at least... I wasn't aware of these limitations.

We have been using VS for Mac so I guess that is why we didn't run into this.

You are right that VS for Mac is full of issues. You don't know from one minute to the next if it is going to suddenly stop working. I had looked into other options as well but they all seem to have issues. Using Hot Restart is also super glitchy and hard to get working. There really is no good solid way of doing iOS work at the moment.

To add fuel to the fire, MS have decided to ditch VS for Mac and they say VS Code is the way to go - but, surprise surprise it is also full of bugs and doesn't work - so literally no way to turn.

I have to say in all my years working in software, this is by far the worst platform I have come across to work with. I have been tearing my hair out and the whole time all MS do is post you tubes with them all having a party showing how amazing Maui is, while at the same time turning a blind eye to all of us 'complainers'...

**Reply by TedMobile** [reply]
Well at least I have some relief that I am not going totally crazy here. I was planning to look at vscode if its a better option but you just nailed it there when you say 

> There really is no good solid way of doing iOS work at the moment.

totally agree with the last paragraph..  not sure where this leaves us but a very uncertain future.

Finally one thing that bothers me  is  why my IPA downloaded to the device still gives me the same issue? I was expecting to just work at runtime when built and deployed.

**Reply by williambuchanan2** [reply]
Another undocumented "feature" is that in entitlements.plist, for debug builds you need to have this for push notifications:
	<key>aps-environment</key>
	<string>development</string>
but that doesn't work in the release builds so before you deploy a release build you need to change it to this:
	<key>aps-environment</key>
	<string>production</string>

**Reply by thomasgalliker** [reply]
What, if you always use aps-environment "production" @williambuchanan2? Also for test apps/backends? We never had problems with that so far (which doesn't mean it's a good solution :))

**Reply by thomasgalliker** [reply]
Question to everyone who was using the old (CrossGeeks) push notitifcation plugin: Do you know what the purpose of this class **DefaultPushNotificationHandler** (Android) ist? It's such an ugly/static/const piece of code, I'm really close to remove it. Do you guys use client-side localization for push notifications? Not really, huh? I mean, you send a notification in a certain language and if the user changes the app language, all old notifications are still in the old language. I don't see a big problem here, do you?

**Reply by williambuchanan2** [reply]
> What, if you always use aps-environment "production"

When we deploy we get this weird app icon, and when I try to run the app I get an error about the app's integrity could not be verified.

![image](https://github.com/dotnet/maui/assets/32165641/f8f5f7bd-9934-4a4c-990d-3d515174562f)

**Reply by williambuchanan2** [reply]
BTW I haven't worked out a way of making this dynamic, so I just need to remember not to check in the entitlements.plist file.

**Reply by TedMobile** [reply]
What we do is when we build the app in azure pipeline we amend the file and change Apns -environment to production and set the task allow to false

**Reply by williambuchanan2** [reply]
> What we do is when we build the app in azure pipeline we amend the file and change Apns -environment to production and set the task allow to false

Not a bad idea...

**Reply by AlleSchonWeg** [reply]
> Another undocumented "feature" is that in entitlements.plist, for debug builds you need to have this for push notifications: aps-environment development but that doesn't work in the release builds so before you deploy a release build you need to change it to this: aps-environment production

I never changed it to production. This should done automatically by xcode, if you compile with release configuration.

**Reply by AlleSchonWeg** [reply]
> Question to everyone who was using the old (CrossGeeks) push notitifcation plugin: Do you know what the purpose of this class **DefaultPushNotificationHandler** (Android) ist? It's such an ugly/static/const piece of code, I'm really close to remove it. Do you guys use client-side localization for push notifications? Not really, huh? I mean, you send a notification in a certain language and if the user changes the app language, all old notifications are still in the old language. I don't see a big problem here, do you?

The android **DefaultPushNotificationHandler** is important, because it shows the notification if you init the plugin( `FirebasePushNotificationManager.Initialize`) without using a custom `IPushNotificationHandler` implementation. 
**DefaultPushNotificationHandler** is also useful, if you want to modify the notification but not want to implement the complete IPushNotificationHandler interface. Thats what i do in my app:
```c#
internal class CustomPushNotificationHandler : DefaultPushNotificationHandler {
	public override void OnReceived(IDictionary<string, object> parameters) {
	  //Do something
	}
	public override void OnBuildNotification(NotificationCompat.Builder notificationBuilder, IDictionary<string, object> parameters) {
     //Do something
	}	
}
```
Init with custom handler:
`FirebasePushNotificationManager.Initialize(this, new CustomPushNotificationHandler(), false, false);`
There are virtual methods you can override like OnBuildNotification which is called just before the notification is displayed. Here you can add a icon for example : `notificationBuilder.SetLargeIcon(...)`

**Reply by thomasgalliker** [reply]
The idea is that you no longer need to call FirebasePushNotificationManager.Initialize(...) because this is already done for you when your MAUI app is started (OnCreate and OnNewIntent are called automatically and the push notification content is forwarded to the plugin automatically).
I'll take your notes into account... thanks for the feedback.

Regarding localization: You translate the push notification in the backend, when sending the notification to FCM, correct? (that's how we do it).

**Reply by AlleSchonWeg** [reply]
@thomasgalliker
Ok. I will try your plugin, if we ready to migrate to maui.
With the CrossGeeks plugin we have to implementent some ugly workarounds to make it working in iOS and Android in all three application states (killed, background, foreground)
Especially this condition in Android: https://github.com/CrossGeeks/FirebasePushNotificationPlugin/blob/d86266a9f45687b418f5f1e69c348681d1ff6e27/Plugin.FirebasePushNotification/DefaultPushNotificationHandler.android.cs#L151
I created a PR at CrossGeeks to make the condition virtual, but this was never merged. Perhaps you could take this into account.

With iOS there are problems with caching the events (OnNotificationOpened, etc.) which makes it nessecary to disable AutoInit.

I translate the notification on the client. The strings are saved in the resx files with a placeholder and the payload contains the data for the placeholder. The string looks like this: "New data available for {0}".
If you do it server side you have to submit and store the current language on a server database. But the user can change the language, And it is not only the translation. Sometimes we have decimal numbers. Depending on the users region the number format is different (decimal separator, etc.). Also date and time formats are different. On iOS we created a NotificationServiceExtension which do the translation stuff. On Android we override the OnReceived method in our CustomPushNotificationHandler:
```c#
		public override void OnReceived(IDictionary<string, object> parameters) {
			if(parameters.TryGetValue("name", out var name) && name != null) {
				parameters.Remove(TitleKey);
				parameters.Remove(BodyKey);
				parameters.Add(TitleKey, AppResource.XXX);
				parameters.Add(BodyKey, string.Format(AppResource.YYY, name));
			}
			base.OnReceived(parameters);
		}
```

**Reply by thomasgalliker** [reply]
Good to know. I can totally understand what you mean with the workarounds. Our code was full of workarounds. I try to bring some fresh air to the plugin. It's good to hear how you use the plugin. I'll come back to you (if you don't mind).

**Comment by TedMobile** [comment]
thanks for all these tips!

**Comment by gabsamples6** [comment]
@thomasgalliker  hi is your push notification plugin still something that you are working on? As you know there isnt much in the maui ecosystem that actually works as push notification is concerned..  thanks in advance
